File: javascript/packages/linter/src/partial-caller-builder.ts
import picomatch from "picomatch"

import { glob } from "tinyglobby"
import { join } from "node:path"
import { readFileSync } from "node:fs"

import { TEMPLATE_GLOB_PATTERN, isPrismNodeType, isRubyRenderLocalNode } from "@herb-tools/core"
import { isOutputRender, renderPartialExpression } from "./rules/prism-rule-utils.js"

import { PartialCallerIndex } from "@herb-tools/core"

import type { ERBRenderNode, HerbBackend, Node, PartialCallSite, PartialIndex, SerializedPartialCallerIndex } from "@herb-tools/core"

const RENDER_MARKER = "render"
const PARSER_OPTIONS = { render_nodes: true, prism_nodes: true } as const

export interface PartialCallerIndexOptions {
  include?: string[]
  exclude?: string[]
}

function templatePatterns(viewRoot: string, include: string[]): string[] {
  const base = viewRoot === "." ? `**/${TEMPLATE_GLOB_PATTERN}` : `${viewRoot}/**/${TEMPLATE_GLOB_PATTERN}`

  return [base, ...include]
}

function templatesIn(projectPath: string, viewRoot: string, include: string[]): Promise<string[]> {
  return glob(templatePatterns(viewRoot, include), { cwd: projectPath, onlyFiles: true, dot: false, absolute: false })
}

function renderNodesIn(node: Node): ERBRenderNode[] {
  const found: ERBRenderNode[] = []

  const walk = (current: Node) => {
    if (current.type === "AST_ERB_RENDER_NODE") found.push(current as ERBRenderNode)

    for (const child of current.childNodes()) {
      if (child) walk(child)
    }
  }

  walk(node)

  return found
}

function localsPassedBy(node: ERBRenderNode): string[] {
  const locals: string[] = []

  for (const local of node.keywords?.locals ?? []) {
    if (!isRubyRenderLocalNode(local)) continue

    const name = local.name?.value

    if (name) locals.push(name)
  }

  return locals
}

function partialNameRenderedBy(node: ERBRenderNode): string | null {
  const call = node.prismNode
  if (!call) return null

  if (!isOutputRender(node)) return null

  const expression = renderPartialExpression(call)
  if (!expression) return null

  if (!isPrismNodeType(expression.node, "StringNode")) return null

  return expression.node.unescaped?.value ?? null
}

export function collectCallSites(herb: HerbBackend, partials: PartialIndex, file: string, source: string, callSites: Map<string, PartialCallSite[]>): number {
  if (!source.includes(RENDER_MARKER)) return 0

  let unresolved = 0

  for (const node of renderNodesIn(herb.parse(source, PARSER_OPTIONS).value)) {
    const name = partialNameRenderedBy(node)

    if (name === null) {
      unresolved++

      continue
    }

    const declaration = partials.lookup(name, file)

    if (!declaration) {
      unresolved++

      continue
    }

    const existing = callSites.get(declaration.file) ?? []

    existing.push({ caller: file, locals: localsPassedBy(node) })
    callSites.set(declaration.file, existing)
  }

  return unresolved
}

export async function buildPartialCallerIndex(herb: HerbBackend, projectPath: string, partials: PartialIndex, options: PartialCallerIndexOptions = {}): Promise<PartialCallerIndex> {
  const files = await templatesIn(projectPath, partials.viewRoot, options.include ?? [])
  const excluded = options.exclude?.length ? picomatch(options.exclude) : null
  const callSites = new Map<string, PartialCallSite[]>()

  let unresolvedRenders = 0
  let skippedFiles = 0

  for (const file of files.sort()) {
    if (excluded?.(file)) {
      skippedFiles++
      continue
    }

    let source: string

    try {
      source = readFileSync(join(projectPath, file), "utf-8")
    } catch {
      continue
    }

    try {
      unresolvedRenders += collectCallSites(herb, partials, file, source, callSites)
    } catch {
      continue
    }
  }

  return new PartialCallerIndex(callSites, unresolvedRenders, skippedFiles)
}

export function partialCallerIndexFrom(data: SerializedPartialCallerIndex | undefined): PartialCallerIndex | undefined {
  return data ? PartialCallerIndex.from(data) : undefined
}

File: rust/herb-linter/src/partial_caller_builder.rs
use std::collections::HashMap;
use std::path::Path;

use crate::utils::erb_utils::is_output_render;
use crate::utils::prism_utils::render_partial_expression;

use herb::action_view_partial_callers::{PartialCallSite, PartialCallerIndex};
use herb::action_view_partial_index::PartialIndex;
use herb::action_view_partial_resolution::TEMPLATE_GLOB_PATTERN;
use herb::nodes::{AnyNode, ERBRenderNode};
use herb::Visitor;

const RENDER_MARKER: &str = "render";
const PROJECT_ROOT: &str = ".";

fn parser_options() -> herb::ParserOptions {
  herb::ParserOptions {
    render_nodes: true,
    prism_nodes: true,
    ..Default::default()
  }
}

fn template_patterns(view_root: &str, include: &[String]) -> Vec<String> {
  let base = if view_root == PROJECT_ROOT {
    format!("**/{TEMPLATE_GLOB_PATTERN}")
  } else {
    format!("{view_root}/**/{TEMPLATE_GLOB_PATTERN}")
  };

  std::iter::once(base).chain(include.iter().cloned()).collect()
}

fn templates_in(project_path: &Path, view_root: &str, include: &[String]) -> Vec<String> {
  herb_config::glob::glob(&template_patterns(view_root, include), project_path, &[] as &[String])
}

#[derive(Default)]
struct RenderCollector {
  renders: Vec<ERBRenderNode>,
}

impl Visitor for RenderCollector {
  fn visit_erb_render_node(&mut self, node: &ERBRenderNode) {
    self.renders.push(node.clone());

    self.walk_erb_render_node(node);
  }
}

fn locals_passed_by(node: &ERBRenderNode) -> Vec<String> {
  let keywords = match node.keywords.as_ref() {
    Some(keywords) => keywords,
    None => return Vec::new(),
  };

  keywords
    .locals
    .iter()
    .filter_map(|local| match local {
      AnyNode::RubyRenderLocalNode(local) => local.name.as_ref().map(|name| name.value.clone()).filter(|name| !name.is_empty()),
      _ => None,
    })
    .collect()
}

fn partial_name_rendered_by(node: &ERBRenderNode) -> Option<String> {
  let call = node.prism()?;

  if !is_output_render(node) {
    return None;
  }

  let expression = render_partial_expression(call)?;

  if !expression.node.is("StringNode") {
    return None;
  }

  expression.node.unescaped.clone()
}

pub fn collect_call_sites(partials: &PartialIndex, file: &str, source: &str, call_sites: &mut HashMap<String, Vec<PartialCallSite>>) -> usize {
  if !source.contains(RENDER_MARKER) {
    return 0;
  }

  let result = match herb::parse_with_options(source, &parser_options()) {
    Ok(result) => result,
    Err(_) => return 0,
  };

  let mut collector = RenderCollector::default();
  collector.visit_document_node(&result.value);

  let mut unresolved = 0;

  for node in &collector.renders {
    let name = match partial_name_rendered_by(node) {
      Some(name) => name,
      None => {
        unresolved += 1;
        continue;
      }
    };

    let declaration = match partials.lookup(&name, Some(file)) {
      Some(declaration) => declaration,
      None => {
        unresolved += 1;
        continue;
      }
    };

    call_sites.entry(declaration.file.clone()).or_default().push(PartialCallSite {
      caller: file.to_string(),
      locals: locals_passed_by(node),
    });
  }

  unresolved
}

#[derive(Default)]
pub struct PartialCallerIndexOptions {
  pub include: Vec<String>,
  pub exclude: Vec<String>,
}

pub fn build_partial_caller_index(project_path: &Path, partials: &PartialIndex, options: &PartialCallerIndexOptions) -> PartialCallerIndex {
  let mut files = templates_in(project_path, &partials.view_root, &options.include);
  let mut call_sites: HashMap<String, Vec<PartialCallSite>> = HashMap::new();

  let mut unresolved_renders = 0;
  let mut skipped_files = 0;

  files.sort();

  for file in files {
    if !options.exclude.is_empty() && herb_config::is_path_matching(&file, &options.exclude) {
      skipped_files += 1;
      continue;
    }

    let source = match std::fs::read_to_string(project_path.join(&file)) {
      Ok(source) => source,
      Err(_) => continue,
    };

    unresolved_renders += collect_call_sites(partials, &file, &source, &mut call_sites);
  }

  PartialCallerIndex::new(call_sites, unresolved_renders, skipped_files)
}
